src/qcmd.1
.\"                                      -*- nroff -*-
.\" qcmd.1 - Manual page for qcmd, qmv and qcp.
.\"
.\" Copyright (C) 2001, 2002, 2004, 2005, 2007, 2008 Oskar Liljeblad
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU Library General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.TH QCMD "1" "December 4, 2007" "qcmd (renameutils)"
.SH NAME
qmv, qcp \- Rename or copy files quickly, editing the file names in a text editor
.SH SYNOPSIS
.B qcp
.RI [ OPTION ]... " " [ FILE ]...

.B qmv
.RI [ OPTION ]... " " [ FILE ]...

.B qcmd
.RI --command= COMMAND " " "" [ OPTION ]... " " [ FILE ]...
.SH DESCRIPTION
This manual page document describes the \fBqcp\fP, \fBqmv\fP and \fBqcmd\fP commands.

The \fBqmv\fP program allows files to be renamed by editing their names
in any text editor. By changing a letter in a text document, a letter
in a filename can be changed. Since the files are listed after each
other, common changes can be made more quickly.

The \fBqcp\fP program works like qmv, but copies files instead of moving
them.

The \fBqmv\fP program was built with safety in mind - bogus renames caused
by human error are avoided by performing as many checks as possible.
Because of this, \fBqmv\fP supports an interactive mode where each step
in the rename procedure can be executed manually. The default is to
start \fBqmv\fP in non-interactive mode. However, when there is a conflict
or error, \fBqmv\fP will drop to interactive mode (instead of losing all
user-made changes). For more information on the renaming process
and interactive mode, see below. This also applies to \fBqcp\fP.

The \fBqcmd\fP program works just like \fBqmv\fP and \fBqcp\fP but allows
you to specify which command to execute.

The \fBls\fP(1) program is used to list files to rename or copy. Therefore
\fBqmv\fP, \fBqcp\fP, and \fBqcmd\fP accepts some \fBls\fP options.
.SH RENAMING/COPYING PROCESS
The process of renaming or copying files consists of many steps. They are:
.TP
.B List files
Generate a list files to rename from command-line arguments
by using \fBls\fP(1).
.TP
.B Create an editable text file
The edit format creates an editable text file with the file names.
.TP
.B Start the editor
Start the text editor and wait until the user has finished editing
it.
.TP
.B Read the edited text file
The edit format reads the edited text file now with updated file names.
.TP
.B Check rename and reorder to resolve conflicts (qmv only)
This is a complicated step which involves the following tasks:

Tag renames where the destination file already exists, renames where
the old file is now missing or inaccessible, and renames
renames where the new name was not changed. Perform a topologic
sort on the renames, so that renaming b\->c, a\->b is possible.
Resolve cross references by renaming into temporary names,
so that renaming e\->f, f\->e (or e\->f, f\->g, g\->e and so on) is possible.

This step results in a plan of renames.
.TP
.B Display the plan.
Display the plan to the user.
.TP
.B Apply the plan.
Apply the plan by actually renaming or copying files (unless \-\-dummy was specified).
.PP
If an error occurs during any of the above steps (except the first),
\fBqmv\fP/\fBqcp\fP drops the user into the interactive mode. This way no changes should
be lost, and errors can be corrected manually before continuing.
See below for a description of the interactive mode.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
.TP
.B \-a, \-\-all
(Passed to \fBls\fP.) Do not hide entries starting with `.'.
.TP
.B \-A, \-\-almost\-all
(Passed to \fBls\fP.) Do not list implied `.' and `..'.
.TP
.B \-B, \-\-ignore\-backups
(Passed to \fBls\fP.) Do not list implied entries ending with `~'.
.TP
.B \-c
(Passed to \fBls\fP.) Sort by ctime (time of last modification).
.TP
.B \-\-command=\fICOMMAND\fR
Execute \fICOMMAND\fR instead of mv or cp.
.TP
.B \-d, \-\-directory
(Passed to \fBls\fP.) List directory entires instead of contents.
.TP
.B \-r, \-\-reverse
(Passed to \fBls\fP.) Reverse order while sorting.
.TP
.B \-R, \-\-recursive
(Passed to \fBls\fP.) List subdirectories recursively.
.TP
.B \-S
(Passed to \fBls\fP.) Sort by file size.
.TP
.B \-\-sort=\fIWORD\fR
(Passed to \fBls\fP.) Sort by extension (\-X), none (\-U), size (\-S),
time (\-t), version (\-v), status (\-c), time (\-t), atime (\-u),
or access (\-u).
.TP
.B \-\-time=\fIWORD\fP
(Passed to \fBls\fP.) If sorting is done by time (\-\-sort=time), sort
by atime, access, use, ctime or status time.
.TP
.B \-t
(Passed to \fBls\fP.) Sort by modification time.
.TP
.B \-u
(Passed to \fBls\fP.) Sort by access time.
.TP
.B \-U
(Passed to \fBls\fP.) Do not sort; list entries in directory order.
.TP
.B \-X
(Passed to \fBls\fP.) Sort alphabetically by entry extension.
.TP
.B \-f, \-\-format=\fIFORMAT\fP
Change edit format of text file. See below for possible values.
.TP
.B \-o, \-\-options=\fIOPTIONS\fP
Pass options to the selected edit format. \fIOPTIONS\fP is in the format

\fIOPTION\fP[=\fIVALUE\fP][,\fIOPTION\fP[=\fIVALUE\fP]...]

For a list of available options for each format, use \-\-options=help
or see below.
.TP
.\".B \-\-show-control-chars
.\"Show non graphic characters as-is when editing renames.
.\".TP
.B \-i, \-\-interactive
Start in command mode (see below for information on this mode).
.TP
.B \-e, \-\-editor=\fIPROGRAM\fP
Specify program to edit text file with. The default program is determined
by looking at the \fBVISUAL\fP environment variable, or if that is not
set, the \fBEDITOR\fP environment variable. If that is not set either,
use the program called \fBeditor\fP.
.TP
.B \-\-ls=\fIPROGRAM\fP
Specify path of the ls program. If you omit the directory, the executable
will be searched for in the directories specified by the PATH environment
variable.
.TP
.B \-v, \-\-verbose
Be more verbose about what is being done.
.TP
.B \-\-dummy
Do everything as usually, except actually renaming any files
("dummy" mode).
.TP
.B \-\-help
Show summary of options.
.TP
.B \-\-version
Output version information and exit.
.SH EDIT FORMATS
An edit format is responsible for generating a text file for editing,
and parsing it once it has been edited. The default format is
dual-column, but there are other formats as well.

Not all edit formats take the same options. Therefore, it is necessary
to specify them using the \-\-options (\-o) option. This option takes a
list of "suboptions" similar to the \-o option in \fBmount\fP(8).

Available edit formats are `single-column' (or `sc'), `dual-column'
(or `dc'), and `destination-only' (or `do'). The default format is
dual-column.
.SH DUAL-COLUMN FORMAT
The \fBdual-column\fP format (`dual-column' or `dc') displays files
in two columns. This is the default and recommended format.
The leftmost column is usually the source file name (which should not
be edited), and the rightmost column the destination file name.
.PP
Supported options:
.TP
.B swap
Swap location of old and new names when editing. I.e. the leftmost
column is now the destination file name, and the rightmost the
source file name.
.TP
.B separate
Put a blank line between all renames.
.TP
.B tabsize=\fISIZE\fP
By default, tab characters of size 8 are used to separate the columns.
With this option the width of these tab characters can be changed.
.TP
.B spaces
Use space characters instead of tab characters when indenting.
.TP
.B width=\fIWIDTH\fP
This option specifies the character position (horizontally) which the
second file name starts at.
.TP
.B autowidth
Normally, if the source file name is longer than \fBwidth\fP characters,
the destination name is printed on the next line instead.
With this option enabled however, qmv/qcp will adjust the width so
that source and destination file names can be displayed on one
line.

If a width has been specified with \fBwidth\fP prior to \fBautowidth\fP,
that width will be used as a minimum width.

Example: Assume that \fBwidth=10,autowidth\fP is specified. Even if all
source file names are shorter than five characters, the split width would
be 10. If there had been one file name longer than 10 characters, the
final width would have been more than 10 characters.

This option is enabled by default.
.TP
.B indicator1=\fITEXT\fP
Text to put before the first file name (column).
.TP
.B indicator2=\fITEXT\fP
Text to put before the second file name (column).
.TP
.B help
Show summary of edit format options.
.SH SINGLE-COLUMN FORMAT
The \fBsingle-column\fP format (`single-column' or `sc') displays
files in a single column - first source file name and on the next
line the destination file name.
.PP
Supported options:
.TP
.B swap
Swap location of old and new names when editing. I.e. the first
line will contain the destination file name, and the next line the
source file name.
.TP
.B separate
Put a blank line between all renames.
.TP
.B indicator1=\fITEXT\fP
Text to put before the first file name.
.TP
.B indicator2=\fITEXT\fP
Text to put before the second file name.
.TP
.B help
Show summary of edit format options.
.SH DESTINATION-ONLY FORMAT
The \fBdestination-only\fP format (`destination-only' or `do') displays
files only the destination file name, one on each line. This format is
generally not recommended, since the only way to identify source file
name is by looking at the line number. But it may be useful with some
text editors.
.PP
Supported options:
.TP
.B separate
Put a blank line between all renames (file names).
.SH INTERACTIVE MODE
In interactive mode \fBqmv\fP/\fBqcp\fP reads commands from the keyboard,
parses them, and executes them. This is done using GNU readline.
.PP
The following commands are available:
.TP
.B ls, list \fR[\fIOPTIONS\fR].. [\fIFILES\fR]..
Select files to rename. If no files are specified, select all files in
current directory. The accepted options are those which are passed to
\fBls\fP(1). Use `help ls' to display a list of these.
.TP
.B import \fIFILE\fR
Read files to rename from a text file. Each line should correspond to an
existing file to rename.
.TP
.B ed, edit
Edit renames in a text editor. If this command has been run before, and
not `all' is specified, only edit renames with errors.
.TP
.B plan
Display the current rename-plan. (This plan is created after `edit'.)
.TP
.B apply
Apply the current plan, i.e. rename files. Only those files marked as OK
in the plan will be renamed.
.TP
.B retry
If some rename failed earlier during `apply', this command will try those
renames again.
.TP
.B show
Display the value of the specified configuration variable, or all
variables if none specified. See below for a list of configuration
variables.
.TP
.B set \fRVARIABLE VALUE
Set the value of a configuration variable.
.TP
.B exit, quit
Exit the program. If there are unapplied changes, the user will
be notified so, and it will be necessary to run this command
an extra time to exit the program.
.TP
.B help \fR[\fIls\fP|\fIusage\fP]
If `ls' is specified, display list options. If `usage' is specified,
display accepted command line options. Otherwise display help on
commands in interactive mode.
.TP
.B version
Display version information.
.SH VARIABLES
The following variables are available in interactive mode:
.TP
.\".B show-control-chars \fIBOOLEAN\fP
.\".TP
.B dummy \fIBOOLEAN\fP
.TP
.B editor \fISTRING\fP
.TP
.B format \fISTRING\fP
.TP
.B options \fISTRING\fP
These variables corresponds to the options with the same name.
.TP
.B tempfile \fISTRING\fP
This variable contains the name of the temporary file which is
edited with `edit'. It cannot be set; only be read with `show'.
.PP
A boolean value is specified as `0', `false', `no', `off'
or `1', `true', `yes', and `on'. Strings are specified without
quotes.
.SH EXAMPLES
Edit names of files in current directory.
.br
	\fBqmv\fP
.PP
Edit names of files with extension `.c'. Sort files by modification time.
.br
	\fBqmv \-t *.c\fP
.PP
Edit names of files using the nedit editor and with column width 100.
.br
	\fBqmv \-enedit \-owidth=100
.SH REPORTING BUGS
Report bugs to <\fIoskar@osk.mine.nu\fP>.
.SH AUTHOR
The author of \fBrenameutils\fP and this manual page is Oskar Liljeblad <\fIoskar@osk.mine.nu\fP>.
.SH COPYRIGHT
Copyright \(co 2001, 2002, 2004, 2005, 2007, 2008, 2011 Oskar Liljeblad

This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


NEWS
2012-04-23: renameutils 0.12.0 released.
  Added `ls' variable and option to specify path of list program
  (suggested by Michael Williams).
  Fix bug with -d option, resulting in "skipping directory argument"
  error (reported by Stroller).
  Link with $(INTLLIBS) to fix Cygwin compilation (patch from Christian
  Franke).
  Fix minor Makefile nit - use $(DESTDIR) (see Gentoo bug #215954).
  Fix Readline history detection problem.
  Updated Gnulib.
  Add note that GNU Coreutils ls is requred (from Michael Williams).

2011-08-18: renameutils 0.11.0 released.
  Added a note about Solaris packages of renameutils on the web page
  (thanks to Steve Christensen).
  Updated French translation from Manuel Pegourie-Gonnard, thanks.
  Applied manpage typo patch from David Sastre, thanks.
  Fix a crash on exit with circular renames (reported by various people).
  Fix bug when a directory would be specified along with files on the
  command line (reported by Axel Beckert through Francois Marier, Debian
  bug #600411).

2008-09-21: renameutils 0.10.0 released.
  Change license to GPL v3.
  Added "import" command, suggested by Lou Afonso.
  Added "--command" option to specify command to use when renaming/copying
  files with qcmd, suggested by Nick Pope and Ulrich Dangel.
  Applied manual page lintian warning patch from Francois Marier.

2007-12-04: renameutils 0.9.0 released.
  Clean debian files in distclean-local, not clean-local.
  Do not remove debian/compat in distclean-local.
  Detect mv and cp options properly for icmd.
  Fix handling of trailing slashes in icmd (reported by Francois Marier).
  Use AC_SYS_LARGEFILE to fix renaming problems with large files (Debian
  bug report).
  Use program_name, not program_invocation_name for improved portability
  (bug reported by Andreas Eriksson).
  Various build and compatibility (Gnulib) fixes.
  Change icmd behavior when multiple arguments are specified - add option
  "--pass-though".
  Miscellaneous icmd improvements. Merge imv and icp manual page.
  Fix `plan' command crash and other crashes, reported by Mike Frysinger.
  Updated French and Swedish translations.

2005-08-23: renameutils 0.8.1 released.
  Now with complete French translation.

2005-08-23: renameutils 0.8.0 released.
  Removed debugging statements rendering imv and icp unusable (oops!)
  (reported by Francois Marier).
  Use mv(1) to perform the actual rename. Fixes Debian bug 316113
  (http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=316113). (Reported by
  Daniel Hulme through Francois Marier.)
  Improved some messages and error messages.
  Change order of inclusion of readline headers, fixing complation problem
  on some distributions (patch from Brian Hall).
  Added --command and --arg-options option to icmd (icp and imv).

2005-08-16: web site updated.
  Add some missing entries to this history.

2005-08-16: renameutils 0.7.0 released
  Added Swedish translation.
  Added French translation (from Francois Marier).
  The executables in the RPM packages now links with ncurses to prevent
  problems on Red Hat and Fedora systems.
  Handle POSIXLY_CORRECT in imv (fixes rename bug) (from Brian Hall).
  Change imv to pass command line through to mv if at least two file names
  are specified (or --target-directory option for mv) (suggested by
  Francois Marier).
  imv now checks if the source file exists prior to asking for destination
  name (suggested by Francois Marier).
  imv now detects and handles options requiring arguments for mv.
  imv now trims the edited file name. This means that the destination file
  name cannot have leading or trailing white space.
  Merge code base for imv and icp into icmd.
  Got rid of year ranges as recommended by the FSF.
  Improve the manual pages a little.
  Add Gnulib modules quote, dirname. Remove stat as it no longer exists in
  Gnulib.
  Fixed some typos in messages of the program.

2005-06-28: web site updated.
  Update description in README and web page.

2005-06-28: renameutils 0.6.0 released.
  Added new programs, icp and qcp (suggested by Mario Vanoni).
  Quote file names and strings before printing them (like GNU coreutils).
  Make use of more gnulib modules (quotearg, yesno, version-etc).
  Fixed a serious memory management-related bug in imv causing renaming to
  fail for no reason.
  Fixed a potential segmentation fault when printing a particular error
  message.
  Added copyright and license header to source files and manual pages.
  Manual page updates:
    Added manual page for deurlname.
    Corrected grammar.
    Removed section in imv about placing it in the Midnight Commander menu.
  Gettext is now external (no intl directory).
  Use Gnulib properly (modules.gnulib and autogen.sh removed).
  Added debian package.
  Added MANIFEST.sources file.
  Fix potential bug in cross-reference solution code.
  Added documentation for qcmd --help from Stephen Kolupaev, thanks.

2004-09-17: renameutils 0.5.1 released.
  Clarified README about requirements.
  Include Makefile.am.head in lib and m4 directories in distribution
  tarball.
  Minor clean ups to the autogen.sh, Makefile.am and configure.ac files.
  Now compiles with --disable-nls as well:
    Don't include (<gettext.h is enough).
    Include <locale.h> where setlocale is used.
  Check that locale was set correctly.

2004-08-26: renameutils 0.5.0 released.
  Added deurlname - a program to remove URL-escaped character sequences
  from filenames (such as %20 for space).
  Internationalized renameutils - added support for gettext.
  Removed debian support since renameutils has been in the official Debian
  archives for quite some time.
  Removed show-control-chars variable and option (automaticly assumed since
  we are using gettext).
  Fixed a bug in dual-column editformat that would cause a segfault when
  renaming files with trailing space in their names (reported by Chris
  Jantzen).
  Fixed a bug in qmv that made it impossible to rename files larger than

src/qcmd.c
/* qcmd.c - Main routines for qcmd.
 *
 * Copyright (C) 2001, 2002, 2004, 2005, 2007, 2008 Oskar Liljeblad
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <config.h>
#include <unistd.h> 	    	    	/* gnulib (POSIX) */
#include <locale.h> 	    	    	/* gnulib (POSIX) */
#include <getopt.h> 	    	    	/* gnulib (POSIX) */
#include <stdlib.h> 	    	    	/* C89 */
#include <errno.h>  	    	    	/* C89 */
#include <string.h>			/* gnulib (C89) */
#include <stdbool.h>	    	    	/* gnulib (POSIX) */
#include <gettext.h>	    	    	/* gnulib (gettext) */
#define _(s) gettext(s)
#define N_(s) (s)
#include "progname.h"	    	    	/* gnulib */
#include "quotearg.h"	    	    	/* gnulib */
#include "version-etc.h"    	    	/* gnulib */
#include "dirname.h"	    	    	/* gnulib */
#include "configmake.h"		   	/* gnulib */
#include "xalloc.h"			/* gnulib */
#include "common/io-utils.h"
#include "common/string-utils.h"
#include "common/error.h"
#include "common/llist.h"
#include "qcmd.h"

char *program;
bool simulate = false;
bool verbose = false;
LList *work_files;
char *work_directory;
char *editor_program = NULL;
char *ls_program = NULL;
char *edit_filename = NULL;
EditFormat *format = &dual_column_format;
ApplyPlan *plan = NULL;
const char version_etc_copyright[] = "Copyright (C) 2001, 2002, 2004, 2005, 2007, 2008, 2011 Oskar Liljeblad";

enum {
    SIMULATE_OPT = 1000,
    COMMAND_OPT,
    LS_OPT,
    VERSION_OPT,
    HELP_OPT,
};

static struct option option_table[] = {
    { "command",      required_argument, NULL, COMMAND_OPT  },
    { "format",       required_argument, NULL, 'f'          },
    { "options",      required_argument, NULL, 'o'          },
    { "editor",	      required_argument, NULL, 'e'          },
    { "ls",           required_argument, NULL, LS_OPT       },
    { "verbose",      no_argument,       NULL, 'v'          },
    { "dummy",        no_argument,       NULL, SIMULATE_OPT },
    { "simulate",     no_argument,       NULL, SIMULATE_OPT },
    { "interactive",  no_argument,       NULL, 'i'          },
    { "version",      no_argument,       NULL, VERSION_OPT  },
    { "help",         no_argument,       NULL, HELP_OPT     },
    { 0 },
};

void
display_version(void)
{
    version_etc(stdout, program, PACKAGE, VERSION, "Oskar Liljeblad", NULL);
}

void
display_help(FILE *out)
{
    fprintf(out, _("Usage: %s [OPTION]... [FILE]...\n\
Move (qmv) or copy (qcp) files quickly, editing the destination file names\n\
in a text editor.\n\
\n"), program_name);
    display_ls_help(out);
    fprintf(out, _("\n\
Format-related options:\n\
  -f, --format=FORMAT        change edit format of text file\n\
  -o, --options=OPTIONS      pass options to the edit format\n\
\n\
Other options:\n\
  -i, --interactive          start in command mode\n\
  -e, --editor=PROGRAM       path of program to edit text file with\n\
      --ls=PROGRAM           path of program to list files with\n\
  -v, --verbose              be more verbose\n\
      --dummy                do not copy (\"dummy\" mode)\n\
\n\
General options:\n\
      --help                 display this help and exit\n\
      --version              output version information and exit\n\
\n\
If no FILE arguments are specified, %s opens a text editor session\n\
showing all files in the current directory. With the default edit format,\n\
the files appear in a single column extending downward. An identical\n\
column appears on its right. Edit the right hand column - make desired\n\
name changes. Save the file and exit. %s will now do a rename/copy\n\
operation for any line changed.\n\
\n\
Unless the --editor option is used, the editor is chosen from the VISUAL\n\
environment variable if set, otherwise EDITOR. If none of these variables\n\
are set, the program `editor' is used.\n\
\n\
Possible values for --format are: `single-column', `dual-column',\n\
and `destination-only'. For a list of available options for each\n\
format, use --options=help.\n"), program, program);
      fprintf(out, _("\nReport bugs to <%s>.\n"), PACKAGE_BUGREPORT);
}

static bool
noncommandmode_run(char **args)
{
    if (!list_files(args))
    	return true;
    if (llist_is_empty(work_files)) {
	warn(_("no files matched"));
	return true;
    }
    if (!edit_files(true, false))
    	return false;
    if (plan != NULL)
	free_plan(plan);
    plan = make_plan(work_files);
    if (plan == NULL)
    	return false;
    display_plan(plan);

    /*dump_spec_list(work_files);*/

    if (!llist_is_empty(plan->error))
    	return false;
    if (!llist_is_empty(plan->ok) && !apply_plan(plan))
    	return false;

    return true;
}

int
main(int argc, char **argv)
{
    struct option *all_options;
    char *format_options = NULL;
    bool interactive = false;

    set_program_name(argv[0]);
    program = base_name(program_name);

    if (setlocale(LC_ALL, "") == NULL)
        warn(_("cannot set locale: %s\n"), errstr);
#ifdef ENABLE_NLS
    if (bindtextdomain(PACKAGE, LOCALEDIR) == NULL)
        warn(_("cannot bind message domain: %s\n"), errstr);
    if (textdomain(PACKAGE) == NULL)
        warn(_("cannot set message domain: %s\n"), errstr);
#endif

    work_directory = xstrdup(".");
    work_files = llist_new();

    /* Initialize option table, and parse options */
    all_options = append_ls_options(option_table);
    for (;;) {
    	int c;

	c = getopt_long(argc, argv, "aABcdrRStuUXvf:o:e:i", option_table, NULL);
	if (c == -1)
	    break;

	switch (c) {
	case '?':
	    exit(1);
	case 'i': /* --interactive */
	    interactive = true;
	    break;
	case 'v': /* --verbose */
	    verbose = true;
	    break;
	case 'f': /* --format */
	    format = find_edit_format_by_name(optarg);
	    if (format == NULL)
		die(_("no such edit format `%s'"), quotearg(optarg));
	    break;
	case 'o': /* options */
	    format_options = optarg;
	    break;
	case 'e': /* --editor */
	    editor_program = optarg;
	    break;
        case LS_OPT: /* --ls */
            ls_program = optarg;
            break;
        case COMMAND_OPT:
            force_command = optarg;
            break;
	case SIMULATE_OPT:
	    simulate = true;
	    break;
	case VERSION_OPT:
	    display_version();
	    exit(0);
	case HELP_OPT:
	    display_help(stdout);
	    exit(0);
	default:
	    process_ls_option(c);
	    break;
	}
    }

    /* Identify program mode - qmv or qcp */
    if (force_command == NULL && strcmp(program, "qmv") != 0 && strcmp(program, "qcp") != 0)
    	die(_("unable to identify mode (start as qmv or qcp)"));

    /* Create a temporary file for editing */
    edit_filename = create_temporary_file(program);
    if (edit_filename == NULL)
	die(_("cannot create temporary file: %s"), errstr);

    /* Find a suitable editor program */
    if (editor_program == NULL
    	    && (editor_program = getenv("VISUAL")) == NULL
    	    && (editor_program = getenv("EDITOR")) == NULL)
    	editor_program = "editor";
    editor_program = xstrdup(editor_program);

    if (ls_program == NULL)
        ls_program = xstrdup("ls");

    /* Parse format options */
    if (format_options != NULL && !format->parse_options(format_options))
	exit(0);

    edit_file_list = llist_new();
    if (interactive) {
	display_commandmode_header();
	if (argv[optind] != NULL)
	    list_files(argv+optind);
	commandmode_mainloop();
    } else {
    	if (!noncommandmode_run(argv+optind)) {
	    printf(_("Entering interactive mode.\n"));
	    commandmode_mainloop();
	}
    }
    llist_free(edit_file_list);

    if (plan != NULL)
	free_plan(plan);

    /*dump_spec_list(work_files);*/
    llist_iterate(work_files, free_file_spec);
    llist_free(work_files);
    free(all_options);
    free(editor_program);
    free(ls_program);
    if (unlink(edit_filename) < 0)
    	warn("cannot delete %s: %s\n", quotearg(edit_filename), errstr);
    free(edit_filename);
    free(work_directory);

    return 0;
}
